[UnifiedPDF] Allow WKImmediateActionController to hit test on UnifiedPDFPlugin

aprotyas opened the pull request:
#### 3ce9b3d1133f25e8fecc81788cded923b4bb7a99
<pre>
[UnifiedPDF] Allow WKImmediateActionController to hit test on UnifiedPDFPlugin
<a href="https://bugs.webkit.org/show_bug.cgi?id=269433">https://bugs.webkit.org/show_bug.cgi?id=269433</a>
<a href="https://rdar.apple.com/122990495">rdar://122990495</a>

Reviewed by Tim Horton.

Any hit testing WebPage::performImmediateActionHitTestAtLocation would
do on the PDF plugin was being compiled away for UnifiedPDFPlugin,
notably because we had not implemented the rectForSelectionInRootView and
lookupTextAtLocation methods on UnifiedPDFPlugin yet.

This patch addresses this issue by implementing the required lookup
methods on UnifiedPDFPlugin and putting the relevant lookup calls behind
ENABLE(PDF_PLUGIN) instead of ENABLE(LEGACY_PDFKIT_PLUGIN) in
performImmediateActionHitTestAtLocation.

We take the liberty to move away from a std::pair to a named struct for
the return type of lookupTextAtLocation, and we also remove its const
qualifier. The latter is necessary so that the plugin can update its
current selection, if required.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::lookupTextAtLocation):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromRootViewToDocument const):
(WebKit::UnifiedPDFPlugin::convertFromPageToRootView const):
(WebKit::UnifiedPDFPlugin::existingSelectionContainsPoint const):
(WebKit::UnifiedPDFPlugin::rectForSelectionInRootView const):
(WebKit::UnifiedPDFPlugin::lookupTextAtLocation):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::lookupTextAtLocation const):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: <a href="https://commits.webkit.org/274728@main">https://commits.webkit.org/274728@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45f44bb919d81d06f4f846393bce791057252d49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18852 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42385 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35743 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16182 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33212 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34446 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13749 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13775 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43664 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36178 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35737 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39513 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12054 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16275 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8928 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16323 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15919 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2248,14 +2250,31 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+bool UnifiedPDFPlugin::existingSelectionContainsPoint(const FloatPoint& rootViewPoint) const
 {
-    return false;
+    auto documentPoint = convertFromPluginToDocument(convertFromRootViewToPlugin(IntPoint { rootViewPoint }));
```
Could make a helper for this

aprotyas replied:
Yeah I should probably just write out convertFromDocumentToRootView (and reverse)

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/45f44bb919d81d06f4f846393bce791057252d49

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2248,14 +2250,31 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+        return { };
+
+    auto rectInPageSpace = IntRect { [selection boundsForPage:page.get()] };
+    IntRect rectInRootView = convertFromPluginToRootView(convertFromDocumentToPlugin(convertFromPageToDocument(rectInPageSpace, *pageIndex)));
```
A helper here would definitely make it easier to read. Probably won't need the local variable either and could return directly since the method name would hopefully make it clear

aprotyas replied:
Also fair, should just probably write out convetFromPageToRootView (and reverse)

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/45f44bb919d81d06f4f846393bce791057252d49

aprotyas committed:
[UnifiedPDF] Allow WKImmediateActionController to hit test on UnifiedPDFPlugin
https://bugs.webkit.org/show_bug.cgi?id=269433
rdar://122990495

Reviewed by Tim Horton.

Any hit testing WebPage::performImmediateActionHitTestAtLocation would
do on the PDF plugin was being compiled away for UnifiedPDFPlugin,
notably because we had not implemented the rectForSelectionInRootView and
lookupTextAtLocation methods on UnifiedPDFPlugin yet.

This patch addresses this issue by implementing the required lookup
methods on UnifiedPDFPlugin and putting the relevant lookup calls behind
ENABLE(PDF_PLUGIN) instead of ENABLE(LEGACY_PDFKIT_PLUGIN) in
performImmediateActionHitTestAtLocation.

We take the liberty to move away from a std::pair to a named struct for
the return type of lookupTextAtLocation, and we also remove its const
qualifier. The latter is necessary so that the plugin can update its
current selection, if required.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::lookupTextAtLocation):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromRootViewToDocument const):
(WebKit::UnifiedPDFPlugin::convertFromPageToRootView const):
(WebKit::UnifiedPDFPlugin::existingSelectionContainsPoint const):
(WebKit::UnifiedPDFPlugin::rectForSelectionInRootView const):
(WebKit::UnifiedPDFPlugin::lookupTextAtLocation):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::lookupTextAtLocation const):
(WebKit:: const): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: https://commits.webkit.org/274728@main
